samcli/lib/build/workflow_config.py

```diff
@@ -2,60 +2,158 @@
 
 
 class UnsupportedRuntimeException(Exception):
     pass
 
 
-def get_workflow_config(runtime):
+def get_workflow_config(runtime, code_dir, project_dir):
     """
-    Get a workflow config that corresponds to the runtime provided
+    Get a workflow config that corresponds to the runtime provided. This method examines contents of the project
+    and code directories to determine the most appropriate workflow for the given runtime. Currently the decision is
+    based on the presence of a supported manifest file. For runtimes that have more than one workflow, we choose a
+    workflow by examining ``code_dir`` followed by ``project_dir`` for presence of a supported manifest.
 
     Parameters
     ----------
     runtime str
         The runtime of the config
 
+    code_dir str
+        Directory where Lambda function code is present
+
+    project_dir str
+        Root of the Serverless application project.
+
     Returns
     -------
     namedtuple(Capability)
         namedtuple that represents the Builder Workflow Config
     """
 
-    workflow_config_by_runtime = {
-        "python2.7": PYTHON_PIP_CONFIG,
-        "python3.6": PYTHON_PIP_CONFIG,
-        "python3.7": PYTHON_PIP_CONFIG,
-        "nodejs4.3": NODEJS_NPM_CONFIG,
-        "nodejs6.10": NODEJS_NPM_CONFIG,
-        "nodejs8.10": NODEJS_NPM_CONFIG,
-        "ruby2.5": RUBY_BUNDLER_CONFIG
+    selectors_by_runtime = {
+        "python2.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.6": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "python3.7": BasicWorkflowSelector(PYTHON_PIP_CONFIG),
+        "nodejs4.3": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs6.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "nodejs8.10": BasicWorkflowSelector(NODEJS_NPM_CONFIG),
+        "ruby2.5": BasicWorkflowSelector(RUBY_BUNDLER_CONFIG),
+
+        # When Maven builder exists, add to this list so we can automatically choose a builder based on the supported
+        # manifest
+        "java8": ManifestWorkflowSelector([
+            # Gradle builder needs custom executable paths to find `gradlew` binary
+            JAVA_GRADLE_CONFIG._replace(executable_search_paths=[code_dir, project_dir])
+        ]),
     }
 
-    try:
-        return workflow_config_by_runtime[runtime]
-    except KeyError:
+    if runtime not in selectors_by_runtime:
         raise UnsupportedRuntimeException("'{}' runtime is not supported".format(runtime))
+
+    selector = selectors_by_runtime[runtime]
+
+    try:
+        config = selector.get_config(code_dir, project_dir)
+        return config
+    except ValueError as ex:
+        raise UnsupportedRuntimeException("Unable to find a supported build workflow for runtime '{}'. Reason: {}"
+                                          .format(runtime, str(ex)))
+
+
+class BasicWorkflowSelector(object):
+    """
+    Basic workflow selector that returns the first available configuration in the given list of configurations
```

Comment:
Is this going to be extended in the future? what is the value of this class, if its just returns the first one?

Reply:
For inheritance.. I didn't create an abstract base class because it didn't seem helpful. But in future we will create a new subclass that will look up Samrc to find the workflow config to use given a runtime. This can be useful for `provided` runtimes or if customer wants to override a default configuration